packages/flutter_tools/lib/src/plugins.dart

```diff
@@ -298,16 +317,67 @@ List<Plugin> findPlugins(FlutterProject project) {
     packages = PackageMap(packagesFile).map;
   } on FormatException catch (e) {
     printTrace('Invalid .packages file: $e');
-    return plugins;
+    return <Plugin>[];
   }
-  packages.forEach((String name, Uri uri) {
-    final Uri packageRoot = uri.resolve('..');
-    final Plugin plugin = _pluginFromPubspec(name, packageRoot);
-    if (plugin != null) {
-      plugins.add(plugin);
+
+  return packages.entries
+    .map((MapEntry<String, Uri> entry) {
+      final Uri packageRoot = entry.value.resolve('..');
+      return _packageFromPubspec(entry.key, packageRoot);
+    })
+    .where((_FoundPackage package) => package != null);
+}
+
+List<Plugin> findPlugins(FlutterProject project) {
+  final Iterable<_FoundPackage> allPackages = _findAllPackages(project);
+  final List<Plugin> allPlugins = allPackages.whereType<Plugin>().toList();
+
+  // if a pubspec file is available, only report plugins that aren't a dev
+  // dependency.
+  final String pubspecFile = fs.path.join(project.directory.path, 'pubspec.yaml');
+  if (!fs.isFileSync(pubspecFile)) {
+    return allPlugins;
+  }
+
+  final dynamic pubspec = loadYaml(fs.file(pubspecFile).readAsStringSync());
+  if (pubspec['dependencies'] == null || pubspec['dependencies'] is! YamlMap) {
+    return allPlugins;
+  }
+
+  final Map<String, _FoundPackage> allPackagesByName = <String, _FoundPackage>{
+    for (final _FoundPackage package in allPackages) package.name: package,
+  };
+
+  // Crawl non-dev packages and their dependencies, starting from the project.
+  // While we're only interested in plugins, we need to crawl all packages
+  // because they could depend on a plugin transitively
+  final YamlMap directDependencies = pubspec['dependencies'] as YamlMap;
+  final Queue<_FoundPackage> pending = ListQueue<_FoundPackage>.from(
```

Comment:
I don't think there is any reason to use a `Queue` here, its implemented with a List anyway. if you want to do a BFS, use a list and pop from the end.